### ISSUE TITLE
Grads are zero in self.s

### ISSUE THREAD

**Issue by PetraZ** [ssui-liu/learnable-embed-sizes-for-RecSys#4]
Hi, I used the PEP module to replace regular Embedding layer in a similar CTR problem with init value -150, and noticed the after the training the self.s value is still all -150, which means it's basically untrained. I did some debugging and notice that almost all grads on self.s are 0 after loss backwards. Indeed with sigmoid(-150), it's becoming 0 in torch. I wonder how you made it trainable with s init value like -150? What's rationale to make the init so small with sigmoid(-150) or sigmoid(-15) ~0? Thanks!

**Comment by PetraZ** [comment]
d sigmoid(x)/ dx = e^x / (e^x + 1)^2 when x = -150, it's about 7.18e-66. I guess that's why it's stuck at it during training and self.s never really updates.

**Comment by ssui-liu** [comment]
Hi,

You're right, the sigmoid gradient at -150 is vanishingly small, preventing updates from the loss. The trick is to use `weight_decay` (e.g., `1e-5`) in your optimizer (like `torch.optim.Adam`).

Decoupled weight decay modifies the gradient update like $g_t \leftarrow g_t + \lambda \theta_{t-1}$. The crucial part is the $+ \lambda \theta_{t-1}$ term (where $\theta_{t-1}$ is your $s$). This term is non-zero even if the gradient from the loss is zero.

This `weight_decay` component effectively nudges $s$away from -150. Once $s$ moves slightly, the sigmoid gradient becomes non-zero, and the standard loss gradients can start updating it properly. The goal is to give the embedding parameters time to be properly optimized by the loss, ensuring they aren't being pruned too early by thresholds $s$.

**Comment by PetraZ** [comment]
Thanks for sharing the tip.

It makes sense that weight decay can help move the init value -150. (Although my attempt with adamW of weight decay 1e5 still doesn't seem to help in my case, maybe I should try other values)

regarding 


> Once s moves slightly, the sigmoid gradient becomes non-zero, and the standard loss gradients can start updating it properly

won't sigmoid gradient still be very small if let's say s moves from -150 -> -149? it has to be until it moves to something like -10 maybe, the gradient would appear?


I have two other questions if you don't mind,

1.  I wonder what does the final self.s value distribution look like in a well trained PEP(assume large embedding)? 
2.  Assume a PEP is well trained, how could we use it to fasten the inference speed. It seems despite of its sparsity, we still need to use the whole embedding * mask. Does this sound right? 
Thanks!
